cmd/outline-ss-server/metrics.go

```diff
@@ -49,9 +62,106 @@ type outlineMetrics struct {
 var _ service.TCPMetrics = (*outlineMetrics)(nil)
 var _ service.UDPMetrics = (*outlineMetrics)(nil)
 
+type ReportTunnelTimeFunc func(IPKey, ipinfo.IPInfo, time.Duration)
+
+type activeClient struct {
+	mu              sync.Mutex
+	IPKey           IPKey
+	clientInfo      ipinfo.IPInfo
+	connectionCount int
+	startTime       time.Time
+}
+
+type IPKey struct {
+	ip        netip.Addr
+	accessKey string
+}
+
+type tunnelTimeTracker struct {
+	mu               sync.Mutex
+	activeClients    map[IPKey]*activeClient
+	reportTunnelTime ReportTunnelTimeFunc
+}
+
+// Reports time connected for all active clients, called at a regular interval.
+func (t *tunnelTimeTracker) reportAll(now time.Time) {
+	if len(t.activeClients) == 0 {
+		logger.Debugf("No active clients. No TunnelTime activity to report.")
+		return
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	for _, c := range t.activeClients {
+		t.reportDuration(c, now)
+	}
+}
+
+// Reports time connected for a given active client.
+func (t *tunnelTimeTracker) reportDuration(c *activeClient, now time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	connDuration := now.Sub(c.startTime)
+	logger.Debugf("Reporting activity for key `%v`, duration: %v", c.IPKey.accessKey, connDuration)
+	t.reportTunnelTime(c.IPKey, c.clientInfo, connDuration)
+	// Reset the start time now that it's been reported.
+	c.startTime = Now()
+}
+
+// Registers a new active connection for a client [net.Addr] and access key.
+func (t *tunnelTimeTracker) startConnection(clientInfo ipinfo.IPInfo, clientAddr net.Addr, accessKey string) {
+	hostname, _, _ := net.SplitHostPort(clientAddr.String())
+	ipKey := IPKey{ip: netip.MustParseAddr(hostname), accessKey: accessKey}
```

Comment:
netip.MustParseAddr can panic. Please don't use "Must" functions.

Also, let's move the IP parsing to the caller and take a netip.Addr instead. No need to be passing unused data in a larger format around. The tigher API makes it clearer what is needed.